File: py/Minimum Path Sum.py
class Solution:
    # @param grid, a list of lists of integers
    # @return an integer
    def minPathSum(self, grid):
        rows = len(grid)
        lines = len(grid[0])
        
        if rows == 0 or lines == 0:
            return 0
        if rows == 1:
            tmp = 0
            for x in grid[0]:
                tmp += x
            return tmp
        
        dp = []  
        for i in range(2):
            dp.append([0]*lines)
        
        tmp = 0
        for i in range(lines):
            tmp += grid[0][i]
            dp[0][i] = tmp
        
        tmp = grid[0][0]
        for i in range(1,rows):
            tmp += grid[i][0]
            dp[i%2][0] = tmp
            for j in range(1,lines):
                dp[i%2][j] = min(dp[(i-1)%2][j],dp[i%2][j-1]) + grid[i][j]
            
        return dp[(rows-1)%2][lines-1]

File: cpp/sum_problem/Minimum Path Sum .cpp
class Solution {
public:
    int min(int a,int b)
    {
        return a > b ? b : a;
    }
    int minPathSum(vector<vector<int> > &grid) {
        int rows = grid.size();
        
        if(rows == 0)
            return 0;
            
        int lines = grid[0].size();
        
        if(lines == 0)
            return 0;
        
        vector<vector<int>> map;
        vector<int> line1(lines);
        vector<int> line2(lines);
        
        map.push_back(line1);
        map.push_back(line2);
        
        int rowssum = 0;
        int linessum = 0;
        
        for(int i = 0;i < lines;i++)
        {
            linessum += grid[0][i];
            map[0][i] = linessum;
        }
        
        rowssum = grid[0][0];
        
        for(int i = 1;i < rows;i++)
        {
            rowssum += grid[i][0];
            map[i % 2][0] = rowssum;
            
            for(int j = 1;j < lines;j ++)
            {
                map[i % 2][j] = min(map[(i+1)%2][j],map[i % 2][j-1]) + grid[i][j]; 
            }
        }
        
        return map[(rows-1)%2][lines-1];
        
    }
};